src/utils/text-measurement.js

```diff
@@ -76,21 +65,20 @@ class TextMeasurement {
    * @return {string} The fitted text.
    */
   getFittedText(text: string, maxWidth: number): string {
-    if (this.minWidth > maxWidth) {
-      return '';
-    }
-    const textWidth = this.getTextWidth(text);
-    if (textWidth < maxWidth) {
+    if (this.getTextWidth(text) < maxWidth) {
       return text;
     }
-    for (let i = 1, len = text.length; i <= len; i++) {
-      const trimmedText = text.substring(0, len - i);
-      const trimmedWidth = this.getTextWidthApprox(trimmedText) + this.minWidth;
-      if (trimmedWidth < maxWidth) {
-        return trimmedText + this.overflowChar;
-      }
+
+    // Approximate the number of characters to truncate to,
+    // using avg character width as reference.
+    const f = (maxWidth - this.minWidth) / this._averageCharWidth;
+    let n = Math.floor(f);
+    if (n === f) {
+      // The approximate width of `n` characters is exactly max width,
+      // so take one character less just in case.
+      n -= 1;
     }
-    return '';
+    return n >= 0 ? text.substring(0, n) + this.overflowChar : '';
```

Comment:
The algorithm seems to yield the same (slightly broken) result than before, but it's much faster because we don't do this loop anymore :-) congrats!

Very likely we should use `n > 0` here so that we don't get results where we see just `...`. What do you think?

Reply:
I'm all for less `fillText` calls.